app/src/processing/app/contrib/ContributionManager.java

```diff
@@ -284,13 +287,118 @@ public boolean accept(File dir, String file) {
+   * anything and providing feedback via the console status area, such as when
+   * the user tries to run a sketch that imports uninstaled libraries.
+   *
+   * @param aList
+   *          The list of AvailableContributions to be downloaded and installed.
+   */
+  public static void downloadAndInstallOnImport(final Base base,
+                                                final ArrayList<AvailableContribution> aList) {
+
+    // To avoid the user from modifying stuff, since this function is only called
+    // during pre-processing
+    base.getActiveEditor().getTextArea().setEditable(false);
+    base.getActiveEditor().getConsole().clear();
+
+    ArrayList<String> installedLibList = new ArrayList<String>();
+
+    // boolean variable to check if previous lib was installed successfully, 
+    // to give the user an idea about progress being made.
+    boolean isPrevDone = false;
+
+    for (AvailableContribution ad : aList) {
+      if (ad.getType() != ContributionType.LIBRARY)
+        continue;
+      try {
+        URL url = new URL(ad.link);
+        String filename = url.getFile();
+        filename = filename.substring(filename.lastIndexOf('/') + 1);
+        try {
+
+          File contribZip = File.createTempFile("download", filename);
+          contribZip.setWritable(true);
+
+          try {
+            // Use the console to let the user know what's happening
+            // The slightly complex if-else is required to let the user know when
+            // one install is completed and the next download has begun without 
+            // interfereing with occur status messages that may arise in the meanwhile
+            String statusMsg = base.getActiveEditor().getStatusMessage();
+            if (isPrevDone)
+              base
+                .getActiveEditor()
+                .statusNotice(statusMsg
+                                + " "
+                                + Language
+                                  .interpolate("contrib.import.progress.download",
+                                               ad.name));
+            else
+              base
+                .getActiveEditor()
+                .statusNotice(Language
+                                .interpolate("contrib.import.progress.download",
+                                             ad.name));
+
+            isPrevDone = false;
+
+            download(url, contribZip, null);
+
+            base.getActiveEditor()
+              .statusNotice(Language
+                              .interpolate("contrib.import.progress.install",
+                                           ad.name));
+            LocalContribution contribution = ad.install(base, contribZip,
+                                                        false, null);
+
+            if (contribution != null) {
+              contribListing.replaceContribution(ad, contribution);
+              if (base.getActiveEditor() != null)
```

Comment:
Please always use { and } for all blocks, no single statements, please. Too prone to errors.